### PR TITLE
drivers: i2c: i2c_gecko: add exclusive access

### LINKED ISSUE #79110: driver: i2c_gecko: add exclusive support
**Is your enhancement proposal related to a problem? Please describe.**
Ensure exclusive access to I2C when using gecko SoCs (silabs).

**Describe the solution you'd like**
Add support for exclusive access

### PR DESCRIPTION
**Pull request by RomainPelletant**
- Add exclusive and atomic access to i2c resource

Fixes #79110

### CODE REVIEW

File: drivers/i2c/i2c_gecko.c [hunk 1]
```diff
@@ -37,6 +37,7 @@ struct i2c_gecko_config {
 
 struct i2c_gecko_data {
 	struct k_sem device_sync_sem;
+	struct k_sem bus_mutex;
```
Comment (teburd) [inline drivers/i2c/i2c_gecko.c:40]: bus_lock might be more appropriate as a name here, as its not a k_mutex

File: drivers/i2c/i2c_gecko.c [hunk 2]
```diff
@@ -75,8 +76,12 @@ static int i2c_gecko_configure(const struct device *dev, uint32_t dev_config_raw
 	i2cInit.master = 0;
 #endif
 
+	k_sem_take(&data->bus_mutex, K_FOREVER);
```
Comment (teburd) [inline drivers/i2c/i2c_gecko.c:79]: the lock doesn't encompass data->dev_config mutation as it likely should

File: drivers/i2c/i2c_gecko.c [hunk 3]
```diff
@@ -121,6 +126,8 @@ static int i2c_gecko_transfer(const struct device *dev, struct i2c_msg *msgs, ui
 			seq.flags |= I2C_FLAG_10BIT_ADDR;
 		}
 
+		k_sem_take(&data->bus_mutex, K_FOREVER);
```
Comment (teburd) [inline drivers/i2c/i2c_gecko.c:129]: The lock should wrap the entirety of transfer, what happens when two threads call transfer simultaneously otherwise?

**Review by teburd — CHANGES_REQUESTED** [review]

File: drivers/i2c/i2c_gecko.c (hunk 1 above)
Comment (RomainPelletant) [reply drivers/i2c/i2c_gecko.c]: bus_lock is more explicite, done

File: drivers/i2c/i2c_gecko.c (hunk 2 above)
Comment (RomainPelletant) [reply drivers/i2c/i2c_gecko.c]: Encompass from data->dev_config

File: drivers/i2c/i2c_gecko.c (hunk 3 above)
Comment (RomainPelletant) [reply drivers/i2c/i2c_gecko.c]: you are right : updated!

File: drivers/i2c/i2c_gecko.c [hunk 4]
```diff
@@ -37,6 +37,7 @@ struct i2c_gecko_config {
 
 struct i2c_gecko_data {
 	struct k_sem device_sync_sem;
+	struct k_sem bus_lock;
```
Comment (jerome-pouiller) [inline drivers/i2c/i2c_gecko.c:40]: I believe it should be a `k_mutex` rather than a `k_sem`.

Comment (RomainPelletant) [reply drivers/i2c/i2c_gecko.c:40]: In my initial commit, the variable was inconsistent (mutex), but the semaphore behavior is correct.
Change variable name from bus_mutex to bus_lock

**Comment by RomainPelletant** [comment]
@jerome-pouiller @teburd no more issues related to I2C access in our project with this fix

**Review by teburd — APPROVED** [review]

**Review by jhedberg — APPROVED** [review]

File: drivers/i2c/i2c_gecko.c (hunk 4 above)
Comment (jhedberg) [reply drivers/i2c/i2c_gecko.c:40]: Yep, and `k_sem` is almost always a better choice than `k_mutex` when it comes to locking. I recommend checking out https://www.youtube.com/watch?v=vla8DTsttyA by @andyross which has lots of interesting insights regarding this.

**Pull request merged**